hibernate-core/src/main/java/org/hibernate/bytecode/internal/bytebuddy/HibernateMethodLookupDispatcher.java

```diff
@@ -83,99 +85,132 @@ static void registerAuthorizedClass(String className) {
 	}
 
 	static {
-		PrivilegedAction<Void> initializeGetCallerClassRequirementsAction = new PrivilegedAction<Void>() {
-
+		// The action below will return the action used at runtime to retrieve the caller stack
+		PrivilegedAction<PrivilegedAction<Class<?>[]>> initializeGetCallerStackAction = new PrivilegedAction<PrivilegedAction<Class<?>[]>>() {
 			@Override
-			public Void run() {
+			public PrivilegedAction<Class<?>[]> run() {
 				Class<?> stackWalkerClass = null;
 				try {
+					// JDK 9 introduced the StackWalker
 					stackWalkerClass = Class.forName( "java.lang.StackWalker" );
 				}
 				catch (ClassNotFoundException e) {
 					// ignore, we will deal with that later.
 				}
 
 				if ( stackWalkerClass != null ) {
+					// We can use a stack walker
 					try {
 						Class<?> optionClass = Class.forName( "java.lang.StackWalker$Option" );
-						stackWalker = stackWalkerClass.getMethod( "getInstance", optionClass )
+						Object stackWalker = stackWalkerClass.getMethod( "getInstance", optionClass )
 								// The first one is RETAIN_CLASS_REFERENCE
 								.invoke( null, optionClass.getEnumConstants()[0] );
 
-						stackWalkerWalkMethod = stackWalkerClass.getMethod( "walk", Function.class );
-						stackFrameGetDeclaringClass = Class.forName( "java.lang.StackWalker$StackFrame" )
+						Method stackWalkerWalkMethod = stackWalkerClass.getMethod( "walk", Function.class );
+						Method  stackFrameGetDeclaringClass = Class.forName( "java.lang.StackWalker$StackFrame" )
 								.getMethod( "getDeclaringClass" );
+						return new StackWalkerGetCallerStackAction(
+								stackWalker, stackWalkerWalkMethod,stackFrameGetDeclaringClass
+						);
 					}
 					catch (Throwable e) {
 						throw new HibernateException( "Unable to initialize the stack walker", e );
 					}
 				}
-
-				return null;
+				else {
+					// We cannot use a stack walker, default to fetching the security manager class context
+					return new SecurityManagerClassContextGetCallerStackAction();
+				}
 			}
 		};
 
-		if ( System.getSecurityManager() != null ) {
-			AccessController.doPrivileged( initializeGetCallerClassRequirementsAction );
+		GET_CALLER_STACK_ACTION = System.getSecurityManager() != null
+				? AccessController.doPrivileged( initializeGetCallerStackAction )
+				: initializeGetCallerStackAction.run();
+	}
+
+	private static Class<?> getCallerClass() {
+		Class<?>[] stackTrace = System.getSecurityManager() != null
+				? AccessController.doPrivileged( GET_CALLER_STACK_ACTION )
+				: GET_CALLER_STACK_ACTION.run();
+
+		// this shouldn't happen but let's be safe
+		if ( stackTrace.length < 4 ) {
```

Comment:
By the way, this could be `3`. Then you should be able to put it in a constant (same for the `3` in the loop below).